### PR TITLE
proxy test failure changes (fixes #266)

### LINKED ISSUE #266: stack test fails sporadically because tests binding the same port run concurrently
I see reproducible proxy test failures with a specific random seed. Other times, the tests pass.

EDIT: Hmm, it's not that clear, fails quite often even without specifying the seed.

```
$ stack test --test-arguments '--seed 182024966'
thentos-adhocracy-0.0.1: test (suite: tests, args: --seed 182024966)
thentos-tests-0.0.1: test (suite: tests, args: --seed 182024966)
Completed all 2 actions.
Test suite failure for package thentos-adhocracy-0.0.1
    tests:  exited with: ExitFailure 1
Full log available at /Users/robert/src/thentos/.stack-work/logs/thentos-adhocracy-0.0.1.log


    
    Thentos.Adhocracy3.Backend.Api.Proxy
      Thentos.Backend.Api.Proxy
        serviceProxy
          an authenticated request
            is proxied
            gets an unchanged body FAILED [1]
            gets the proxy path added FAILED [2]
    Thentos.Adhocracy3.Backend.Api.Simple
      Thentos.Backend.Api.Adhocracy3
        A3UserNoPass
          has invertible *JSON instances
        A3UserWithPassword
          has invertible *JSON instances
          rejects short passwords
          rejects long passwords
          rejects invalid email addresses
          rejects empty user names
          rejects user names with @
          rejects user names with too much whitespace
    
    Failures:
    
      tests/Thentos/Adhocracy3/Backend/Api/ProxySpec.hs:66: (best-effort)
      1) Thentos.Adhocracy3.Backend.Api.Proxy.Thentos.Backend.Api.Proxy.serviceProxy, an authenticated request, gets an unchanged body
           uncaught exception: HttpException (StatusCodeException (Status {statusCode = 404, statusMessage = "Not Found"}) [("Transfer-Encoding","chunked"),("Date","Tue, 28 Jul 2015 15:50:46 GMT"),("Server","Warp/3.0.13.1"),("X-Response-Body-Start","not found"),("X-Request-URL","POST http://localhost:7118/%18%C3%BF")] (CJ {expose = []})) (after 5 tests and 4 shrinks)
           *** Failed! Exception: 'StatusCodeException (Status {statusCode = 404, statusMessage = "Not Found"}) [("Transfer-Encoding","chunked"),("Date","Tue, 28 Jul 2015 15:50:46 GMT"),("Server","Warp/3.0.13.1"),("X-Response-Body-Start","not found"),("X-Request-URL","POST http://localhost:7118/")] (CJ {expose = []})' (after 5 tests and 4 shrinks): 
           ""
           NonEmpty {getNonEmpty = "a"}
    
      tests/Thentos/Adhocracy3/Backend/Api/ProxySpec.hs:73: (best-effort)
      2) Thentos.Adhocracy3.Backend.Api.Proxy.Thentos.Backend.Api.Proxy.serviceProxy, an authenticated request, gets the proxy path added
           uncaught exception: HttpException (StatusCodeException (Status {statusCode = 404, statusMessage = "Not Found"}) [("Transfer-Encoding","chunked"),("Date","Tue, 28 Jul 2015 15:50:46 GMT"),("Server","Warp/3.0.13.1"),("Cache-Control","no-cache, no-store, must-revalidate"),("Expires","0"),("X-Response-Body-Start","not found"),("X-Request-URL","GET http://localhost:7118/")] (CJ {expose = []})) (after 1 test)
           ""
    
    Source locations marked with "best-effort" are calculated heuristically and may be incorrect.
    
    Randomized with seed 182024966
    
    Finished in 7.0096 seconds
    11 examples, 2 failures

```

### PR DESCRIPTION
**Pull request by robx**
This makes #266 fail more usefully for me. It also seems to head towards something more correct, but is clearly not the right solution in this state.

### CODE REVIEW

**Comment by robx** [comment]
By now, this fixes the issue, in a rather minimal way. Summary:
* bind to localhost instead of `*`, making the attempt binding port 7118 concurrently with other tests fail
* attempt to capture exceptions in the background tasks and make these cause the tests to fail explicitly (the async/link/wait dance seems to work, but I'm not at all sure this is the right way to do it)
* make the proxy bind to a random open port, fixing the actual conflict (the destination port is not as easy to change, because that's part of the backend config as far as I understand)

File: thentos-tests/src/Thentos/Test/Network.hs [hunk 1]
```diff
@@ -0,0 +1,14 @@
+module Thentos.Test.Network
+where
+
+import Network.Socket --(bind, listen, socket, Family(AF_INET), SocketType(Stream))
```
Comment (fhartwig) [inline thentos-tests/src/Thentos/Test/Network.hs]: This needs fixing.

**Comment by fhartwig** [comment]
Looks good: :+1: One nitpick, but I've fixed this myself rather than waiting for you to be at the office. There may be nicer solutions than async exceptions, but this works and is definitely an improvement over the status quo.

**Comment by robx** [comment]
Thanks, I had lost sight of that import TODO.

**Pull request merged**